apps/common/lib/lexical/ast.ex

```diff
@@ -187,9 +187,40 @@ defmodule Lexical.Ast do
   end
 
   @doc """
-  Returns the path to the cursor in the given document at a position.
+  Returns the path to the innermost node in the document at the given position.
 
-  May return a path even in the event of syntax errors.
+  This function differs from `cursor_path/2` in that it expects a valid
+  AST and the returned path will not contain a `:__cursor__` node.
+  """
+  @spec path_at(Document.t(), Position.t()) ::
+          {:ok, [Macro.t(), ...]} | {:error, :not_found | parse_error()}
+  @spec path_at(Macro.t(), Position.t()) ::
+          {:ok, [Macro.t(), ...]} | {:error, :not_found}
+  def path_at(%Document{} = document, %Position{} = position) do
+    with {:ok, ast} <- from(document) do
+      path_at(ast, position)
+    end
+  end
+
+  def path_at(ast, %Position{} = position) do
+    path =
+      Future.Macro.path(ast, fn node ->
+        terminal_node?(node) and contains_position?(node, position)
+      end)
+
+    case path do
+      nil -> {:error, :not_found}
+      path -> {:ok, path}
+    end
```

Comment:
The reason I used `{:error, :not_found}` is because it's also possible that this function returns a syntax error from parsing, which is also a 2-element `:error` tuple, and it seems a little icky for a function to return both `:error` and `{:error, something}`.

Reply:
i see.